compiler/noirc_evaluator/src/ssa/acir_gen/acir_ir/acir_variable.rs

```diff
@@ -670,9 +678,27 @@ impl<F: AcirField> AcirContext<F> {
     pub(crate) fn add_var(&mut self, lhs: AcirVar, rhs: AcirVar) -> Result<AcirVar, RuntimeError> {
         let lhs_expr = self.var_to_expression(lhs)?;
         let rhs_expr = self.var_to_expression(rhs)?;
-        let sum_expr = &lhs_expr + &rhs_expr;
 
-        Ok(self.add_data(AcirVarData::from(sum_expr)))
+        let mut sum_expr = &lhs_expr + &rhs_expr;
+        let sum_expr_fits = fits_in_one_identity(&sum_expr, self.expression_width);
+
+        if !sum_expr_fits {
+            let lhs_witness_var = self.get_or_create_witness_var(lhs)?;
+            let lhs_witness_expr = self.var_to_expression(lhs_witness_var)?;
+            let rhs_witness_var = self.get_or_create_witness_var(rhs)?;
+            let rhs_witness_expr = self.var_to_expression(rhs_witness_var)?;
+
+            let new_sum_expr = &lhs_witness_expr + &rhs_witness_expr;
+            let sum_var = self.add_data(AcirVarData::from(new_sum_expr.clone()));
+
+            // Create new witness for the sum expression
+            let sum_witness_var = self.get_or_create_witness_var(sum_var)?;
+            sum_expr = self.var_to_expression(sum_witness_var)?;
+        }
+
+        let sum_var = self.add_data(AcirVarData::from(sum_expr));
+
+        Ok(sum_var)
```

Comment:
I'm a bit leery of enforcing that the output is a witness also, Off the top of my head this seems like it would be overly aggressive in the majority of cases. Are there good examples of where this helps?

We can clean this function up by reentering after squashing the inputs.

```suggestion
        let sum_expr = &lhs_expr + &rhs_expr;
        let sum_expr_fits = fits_in_one_identity(&sum_expr, self.expression_width);

        if sum_expr_fits {
                let sum_var = self.add_data(AcirVarData::from(sum_expr));

               Ok(sum_var)
        } else {
            // Expression has overflowed maximum size so we simplify `lhs` and `rhs` into witnesses.
            let lhs_witness_var = self.get_or_create_witness_var(lhs)?;
            let rhs_witness_var = self.get_or_create_witness_var(rhs)?;
            
            let sum_var = self.add_var(lhs_witness_var, rhs_witness_var);
            // Note that we also force the output variable to be assigned to a witness.
            self.get_or_create_witness_var(sum_var)
        }
```

Reply:
I also think we are not greedy enough here.
I would do something like:
get the operand that has the biggest expression width, creates a witness out of it and see if the new expression with the witness has proper width. If not, then creates a witness of the other one.
Only if the result expression width is equal to the target width, then I would create a witness of the result.

The other problem is when we create a witness of an operand, this operand could have been used somewhere else, not benefiting from its witness form. This is a bit annoying but not easy to solve.
One idea, when we do not fit, is to analyse the operands and see if one of them has such ancestor (i.e an expression that has been transformed into a witness).

Reply:
> Off the top of my head this seems like it would be overly aggressive in the majority of cases.

Yeah this was overly aggressive in the majority of cases. I was mainly going off of how to optimize this logic:
```rust
fn oh_no_not_again(mut state: [Field; 2]) -> [Field; 2] {
    for _ in 0..200 {
        state[0] = state[0] * state[0] + state[1];
        state[1] += state[0];
    }
    state
}
```
However, that is pretty limiting. This was the main case where I saw a big reduction when creating a witness for the sum expression result. 

> get the operand that has the biggest expression width, creates a witness out of it and see if the new expression with the witness has proper width. If not, then creates a witness of the other one.

I have done this now and saw improvements to poseidon over the very greedy original appraoch. I had originally attempted to do something like this but was getting worse numbers so that is why I went for the very greedy approach. I must have had incorrect logic previously as it looks to be better now 🤷 

> Only if the result expression width is equal to the target width, then I would create a witness of the result.

I experimented with this but did not see any noticeable improvements. In fact for the loop posted above when I create a witness of the sum result each time, I got 301 acir gates which is line with the optimized solution. However, this added more gates for poseidon. So I went with not creating a witness of the result.

Reply:
> The other problem is when we create a witness of an operand, this operand could have been used somewhere else, not benefiting from its witness form. This is a bit annoying and not easy to solve.
> One idea, when we do not fit, is to analyse the operands and see if one of them has such ancestor (i.e an expression that has been transformed into a witness).

Good note. Now that we are moving to place this logic under a flag, we can experiment with this in a follow-up PR